CenterForOpenScience/osf.io
# Feature/Support new preprint versions with different extensions [ENG-148]

## Purpose

Support upload preprint versions being uploaded with different extensions.  Part of this involves storing a history of the file's name across different versions, so earlier versions can still be downloaded correctly.

- ability to upload a preprint primary file with a name that differs from the originally uploaded file (including extension)
- ability to see a history of all versions of the preprint primary file, even across file name and/or extension changes
- ability to download any previous version of the preprint primary file and have it be named correctly (with the correct extension, at a minimum)

## Changes

Important to note, we can't store the name on the version object, because files and versions are a many-to-many relationship. Files can have many versions and versions can have many files.  We create a custom through table and store the name on this table.

- Create a custom through table to link files and versions (currently we're using the automatically-created basefilenode_version table.  Added a osf_basefileversionsthrough table to use instead - columns are file, version, and version_name.
- Migrations added to create the new through table, move data from the automatic through table to the custom one, plus populating the name. 
- Since we're using a custom through table, you can't do things like file.versions.add(version) any longer.  You have to use the custom through table.  Method added to BaseFileNode to take care of adding versions.
- Method added to Version object to retrieve the link from the through table (a specific file linked with a specific version with the name attached)
- Moving/copying behavior modified so when a version is created, the current version name is attached to the file-version link.
- Renaming a file both renames the file object and the file-version object stored in the through table.
- Version name displayed in API - this is pulled from the through table.
- Migration prep - create an assortment of files, add versions to them. Assert that files and versions all there post-migration.

## Example
### Old through table: basefilenode_versions. 
Stores which files belong to which versions and vice versa.  Just a table full of file ids and version ids. Automatically created for us in Django:
<img width="506" alt="Screen Shot 2019-04-08 at 10 52 59 AM" src="https://user-images.githubusercontent.com/9755598/55738399-b0826400-59ec-11e9-8898-d0afa37e038e.png">

### New through table : basefileversionsthrough
Still links files to versions, but stores the name as well.   Since files/versions are a many-to-many relationship, storing the name here helps us retain which name that version had for that particular file.
<img width="874" alt="Screen Shot 2019-04-08 at 10 53 04 AM" src="https://user-images.githubusercontent.com/9755598/55738542-ee7f8800-59ec-11e9-92b3-c835c5b86120.png">


### Name added to file version API - is the unique name stored with the file-version-link
<img width="954" alt="Screen Shot 2019-04-08 at 10 58 33 AM" src="https://user-images.githubusercontent.com/9755598/55738749-5d5ce100-59ed-11e9-9760-fa8238a7a040.png">

## QA Notes
- Renaming a file needs to also rename its fileversion.   Earlier versions need to keep old names.  To test, 1) Upload a file to a project 2) Upload a new version of that file 3) Rename the file 4) View the file's versions in the API: http://localhost:8000/v2/files/{file_id}/versions/.   Version 1 should have the original name, version 2 should display the new name. 
-  Copying files needs to retain old version names.  To test, 1) Upload a file to a project 2) Upload new versions of this file.  Rename the file.  3) Copy that file to a different folder/different directory.  4) View the copied file in the API, and verify that the version names are correct (they should be different)
- This needs thorough regressions of file operations.  Uploading, deleting, moving, copying, adding new versions, renaming, downloading.   Test osfstorage and a couple of different addons.
- Test legacy file revisions page - test that download old revisions works as expected. If older file versions were associated with different names, the older version file download name should now include the old name plus the date.  If you can download revisions of addon files, test that too.
- Double check file operations/file revisions page for a couple of addons.
- File revisions download should retain old name in older versions.
## Documentation



## Side Effects

- No fangorn changes here, you still can't upload a new version that has a new extension in fangorn.  This is going to be implemented for preprints first. 
## Ticket

https://openscience.atlassian.net/browse/ENG-148
